src/frontend/src/session.rs

```diff
@@ -382,6 +382,40 @@ impl FrontendEnv {
         });
         join_handles.push(join_handle);
 
+        // Heartbeat
+        let heartbeat_worker_node_manager = worker_node_manager.clone();
+        let join_handle = tokio::spawn(async move {
+            let duration = std::cmp::max(
+                Duration::from_secs(meta_config.max_heartbeat_interval_secs as _) / 10,
+                Duration::from_secs(1),
+            );
+
+            let mut check_heartbeat_interval =
+                tokio::time::interval(core::time::Duration::from_secs(2));
+            check_heartbeat_interval
+                .set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);
+            check_heartbeat_interval.reset();
+            loop {
+                check_heartbeat_interval.tick().await;
+                for worker in heartbeat_worker_node_manager
+                    .list_worker_nodes()
+                    .into_iter()
+                    .filter(|w| w.property.as_ref().map_or(false, |p| p.is_serving))
```

Comment:
Should we also take streaming node into consideration?

Reply:
The property `is_serving` means whether it could be used as a serving role. Normally, compute nodes have `is_streaming = true` and `is_serving = true` properties.

Reply:
Makes sense. The frontend seems to never communicate with a streaming-only compute node.